units_cpp/game_state.cc

```diff
@@ -1,240 +1,350 @@
+		return winners;
+	}
+	else if (winners.size() == 0) {
+		// No generals remain - the game is a draw. Return all players' numbers.
+		for (size_t i = 0; i < players.size(); i++) {
+			winners.push_back(i);
+		}
+		return winners;
+	} else {
+		// More than one general still remains - the game is still ongoing. Return an empty vector.
+		return std::vector<int>();
+	}
 }
 
+
 const GameRules &GameState::get_rules() const {
-  return rules;
+	return rules;
 }
 
+
 const std::vector<Player> &GameState::get_players() const {
-  return players;
+	return players;
 }
 
+
 const std::vector<std::vector<std::vector<Action>>> &GameState::get_action_log() const {
-  return action_log;
+	return action_log;
 }
 
+
 const std::vector<std::shared_ptr<Battle>> &GameState::get_battles() const {
-  return battles;
+	return battles;
 }
 
+
 int GameState::get_turn() const {
-  return turn;
+	return turn;
+}
+
+
+//------------------------------------------------------------------------------
+// OTHER METHODS
+//------------------------------------------------------------------------------
+void to_json(json &j, const GameState &game_state) {
+	j["players"] = game_state.players;
+	j["action_log"] = game_state.action_log;
+	for (const auto battle : game_state.battles) {
+		j["battles"].push_back(*battle);
+	}
+	j["turn"] = game_state.turn;
+	j["players_ready"] = game_state.players_ready;
+}
+
+
+void GameState::begin_turn() {
+	/* Advance the turn counter, and request all players begin their turns. */
+
+	turn++;
+	players_ready = 0;
+
+	// Request all players begin their turn
+	for (auto &player : players) {
+		player.begin_turn();
+	}
+
+	// Prepare a new vector of actions for each player, for the current turn.
+	action_log.push_back(std::vector<std::vector<Action>>(players.size()));
+}
+
+
+void GameState::execute_battle() {
+	/* Creates a new Battle object and adds it to the record. */
+
+	Battle *b = new Battle(players);
+	battles.push_back(std::shared_ptr<Battle>(b));
+
+	// Do all the damage from the attacks
+	for (auto &atk : b->get_attacks()) {
+		Unit &unit = players[atk.to_player].units.at(atk.to_unit);
+		unit.shields -= atk.damage;
+		if (unit.shields < 0) {
+			unit.health += unit.shields;
+			unit.shields = 0;
+		}
+	}
+	b->set_players_after(players);  // Update players_after
+
+	// Kill dead units, restore shields for alive units
+	for (auto &player : players) {
+		for (auto it = player.units.begin(); it != player.units.end();) {
+			if (it->second.health <= 0) {
+				auto to_delete = it;
+				it++;
+				player.units.erase(to_delete);
+			}
+			else {
+				it->second.shields = it->second.kind.get_shields();
+				it++;
+			}
+		}
+	}
+	if (get_winners().size() == 0) {
+		begin_turn();
+	}
+}
+
+
+bool GameState::perform_action(int player, const Action &action) {
+	if (get_winners().size() > 0) {
+		// Game is already over - all actions illegal
+		return false;
+	}
+	if ((size_t)player >= players.size()) {
+		// Nonexistent player
```

Comment:
Thanks for interpreting all these checks in plain English! I admit I was getting a bit tired of re-understanding them each time myself. I think splitting the single-line checks into three lines w/ an open/closing curly brace is a bit overkill and I will be reverting those, but I'll keep all your comments.